ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -870,4 +802,132 @@ long getAllocatedUsableCapacity() {
+      // call. Therefore, when the call to add a listener returns, the initial notification will have been
+      // received and handled.
+      manager.addInstanceConfigChangeListener(clusterChangeHandler);
+      logger.info("Registered instance config change listeners for Helix manager at {}", zkConnectStr);
+      manager.addIdealStateChangeListener(clusterChangeHandler);
+      logger.info("Registered ideal state change listeners for Helix manager at {}", zkConnectStr);
+      // Now register listeners to get notified on live instance change in every datacenter.
+      manager.addLiveInstanceChangeListener(clusterChangeHandler);
+      logger.info("Registered live instance change listeners for Helix manager at {}", zkConnectStr);
+
+      // in case initial event occurs before adding routing table listener, here we explicitly set snapshot in
+      // ClusterChangeHandler. The reason is, if listener missed initial event, snapshot inside routing table
+      // provider should be already populated.
+      clusterChangeHandler.setRoutingTableSnapshot(routingTableProvider.getRoutingTableSnapshot());
+      // the initial routing table change should populate the instanceConfigs. If it's empty that means initial
+      // change didn't come and thread should wait on the init latch to ensure routing table snapshot is non-empty
+      if (clusterChangeHandler.getRoutingTableSnapshot().getInstanceConfigs().isEmpty()) {
+        // Periodic refresh in routing table provider is enabled by default. In worst case, routerUpdater should
+        // trigger routing table change within 5 minutes
+        logger.info("Routing table snapshot in {} is currently empty. Waiting for initial notification.", dcName);
+        clusterChangeHandler.waitForInitNotification();
+      }
+
+      if (!clusterMapConfig.clustermapListenCrossColo && manager != localManager) {
+        manager.disconnect();
+        logger.info("Stopped listening to cross colo ZK server {}", zkConnectStr);
+      }
+    }
+
+    /**
+     * Currently, this does not connect to the VCR zookeeper and assumes that all partitions are supported in the cloud
+     * datacenter. This will be the case until the VCR and native storage clusters are unified under the same
+     * {@link ClusterMap}. Once this happens, we can use the VCR cluster as a source of truth for supported partitions
+     * in the cloud datacenter.
+     */
+    private void initializeCloudDatacenter() {
+      DcInfo dcInfo = new DcInfo(dcName, dcZkInfo, null, null);
+      dcToDcInfo.put(dcName, dcInfo);
+      dcIdToDcName.put(dcZkInfo.getDcId(), dcName);
+      CloudChangeListener listener =
+          new CloudChangeListener(dcName, clusterMapConfig, clusterChangeHandlerCallback);
+      registerClusterMapListener(listener);
+      // seed existing replicas.
+      for (Set<AmbryReplica> replicas : ambryPartitionToAmbryReplicas.values()) {
```

Comment:
I don't quite understand this part. Will talk with you offline.

Reply:
This logic will have to change a bit since I am rebasing on your changes to listening. However, the main goal is to add any ambry partitions that already exist.

I am roughly trying to accomplish what ReplicationManager does where it first adds the listener, and then iterates through all existing replicas right afterwards so that it knows about all replicas added before the listener was added.

I may just make a new listener for just newly added partitions for this use case since I envision it would be somewhat temporary, as eventually ClusterMap will get the set of managed partitions from the VCR.